File: src/test/java/com/example/coupon/controller/CouponController.java
"redeemCoupon should return OK when coupon is successfully redeemed" {
        // Given:
        val couponId: Long = 123
        val couponService = mockk<CouponService>()
        val couponController = CouponController(couponService)

        every { couponService.redeemCoupon(couponId) } returns true

        // When:
        val response: ResponseEntity<String> = couponController.redeemCoupon(couponId)

        // Then:
        response.statusCode shouldBe HttpStatus.OK
        response.body shouldBe "쿠폰 사용 성공"
    }

    "redeemCoupon should return BadRequest when no coupons are left" {
        // Given:
        val couponId: Long = 456
        val couponService = mockk<CouponService>()
        val couponController = CouponController(couponService)

        every { couponService.redeemCoupon(couponId) } returns false

        // When:
        val response: ResponseEntity<String> = couponController.redeemCoupon(couponId)

        // Then:
        response.statusCode shouldBe HttpStatus.BAD_REQUEST
        response.body shouldBe "잔여 쿠폰 없음"
    }

    "createCoupon should return OK when coupon is successfully created" {
        // Given:
        val couponCreate = CouponCreate("SummerSale", 10, 100L)
        val couponService = mockk<CouponService>(relaxed = true) // relaxed = true for void methods
        val couponController = CouponController(couponService)

        every { couponService.create(couponCreate) } returns Unit

        // When:
        val response: ResponseEntity<String> = couponController.createCoupon(couponCreate)

        // Then:
        response.statusCode shouldBe HttpStatus.OK
        response.body shouldBe "쿠폰 생성 성공"
    }
})

File: src/test/java/com/example/coupon/controller/CouponController.kt
class CouponControllerTest : FunSpec({

    val couponService = mockk<CouponService>()
    val couponController = CouponController(couponService)

    test("redeemCoupon - Successful redemption") {
        // Given: A coupon ID and the coupon service returns true for successful redemption
        val couponId: Long = 123L
        every { couponService.redeemCoupon(couponId) } returns true

        // When: redeemCoupon method is called
        val response: ResponseEntity<String> = couponController.redeemCoupon(couponId)

        // Then: The response should be OK and the body should contain success message
        response.statusCode shouldBe HttpStatus.OK
        response.body shouldBe "쿠폰 사용 성공"
    }

    test("redeemCoupon - Unsuccessful redemption") {
        // Given: A coupon ID and the coupon service returns false for unsuccessful redemption (no remaining coupons)
        val couponId: Long = 456L
        every { couponService.redeemCoupon(couponId) } returns false

        // When: redeemCoupon method is called
        val response: ResponseEntity<String> = couponController.redeemCoupon(couponId)

        // Then: The response should be Bad Request and the body should contain an appropriate error message
        response.statusCode shouldBe HttpStatus.BAD_REQUEST
        response.body shouldBe "잔여 쿠폰 없음"
    }

    test("createCoupon - Successful coupon creation") {
        // Given: CouponCreate data and the coupon service create method is mocked
        val couponCreate = CouponCreate("TestCoupon", 10, 100L)
        every { couponService.create(couponCreate) } returns Unit // void method is mocked as Unit

        // When: createCoupon method is called
        val response: ResponseEntity<String> = couponController.createCoupon(couponCreate)

        // Then: The response should be OK and the body should contain success message
        response.statusCode shouldBe HttpStatus.OK
        response.body shouldBe "쿠폰 생성 성공"
    }

})